File: golang/topkfrequentelements347/top-k-frequent-elements-347.go
package main

import "fmt"

func main() {
	nums := [...]int{3, 3, 2, 0, 0, 2, 4, 4, 5, 5, 9, 9, 9}
	fmt.Println(topKFrequent(nums[:], 1))
	fmt.Println("HFAFASDA")
}

func topKFrequent(nums []int, k int) []int {
	m := make(map[int]int)
	s := make([][]int, len(nums)+1)
	for i := range s {
		s[i] = make([]int, 0)
	}
	for _, v := range nums {
		if _, ok := m[v]; ok {
			m[v] = m[v] + 1
		} else {
			m[v] = 1
		}
	}

	for k, v := range m {
		s[v] = append(s[v], k)
	}
	res := make([]int, 0)

	f := len(nums)

	for f > 0 && len(res) < k {
		if len(s[f]) > 0 {
			for _, v := range s[f] {
				res = append(res, v)
				if len(res) >= k {
					break
				}
			}
		}
		f--
	}
	return res
}

File: nodejs/top-k-frequent-elements-347.js
'use strict';
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
    let freqs = new Map();
    let buckets = [];
    for (let i = 0; i <= nums.length; i++) {
        buckets[i] = new Set();
    }

    for (let v of nums) {
        buckets[0].add(v);
        freqs[v] = 0;
    }

    for (let v of nums) {
        buckets[freqs[v]].delete(v);
        buckets[freqs[v] + 1].add(v);
        freqs[v]++;

    }

    console.log(buckets);
    let res = [];

    let f = buckets.length - 1;
    while (res.length < k && f >= 0) {
        if (buckets[f].size > 0) {
            for (let v of buckets[f]) {
                if (res.length < k) {
                    res.push(v);
                }
            }
        }
        f--;
    }
    return res;
};

console.log(topKFrequent([1,2,1,1,1,1], 3));